apps/labs/posts/automated-accessibility-testing-with-bokeh.md

```diff
@@ -0,0 +1,333 @@
+
+A testing environment needs tests, and to test what we'd done, we needed a
+variety of valid data visualizations. Luckily, Bokeh very generously had over
+100 example plots in the documentation that we could test against. A caveat to
+testing against the documentation was the fact that it would always need to be
+built and served locally before we could run axe-core tests against it.
+
+When the internship started, we had the following
+[project](https://github.com/Quansight-Labs/open-source-internships/issues/31)
+outcomes:
+
+1. Familiarizing with Bokeh concepts.
+2. Initial implementation of automated testing with Playwright and Axe-core.
+3. Building on existing tests.
+4. Integrating tests with Bokeh's CI.
+5. Adding relevant test documentation.
+
+There were a few limitations to achieving all these objectives. The limitations
+were:
+
+- Complexities of the Bokeh codebase that required weeks of knowledge share
+  sessions.
+- The `<canvas>` element being something that couldn't properly be tested.
+- Initial difficulty in getting Playwright and Axe-core to run in the tests.
+- Time constraints.
+
+We were, however, able to achieve the following:
+
+- An audit and
+  [summary](https://github.com/bokeh/bokeh/discussions/14057#discussioncomment-10674506)
+  of ~113 doc examples that were run through axe-core.
+- Familiarity with the Bokeh codebase.
+- An implementation of tests that support Playwright and Axe-core.
+- A provision for creating new integration and unit tests and marking as
+  "accessibility-focused".
+
+Overall, our efforts towards going from 0 to 1 ended up looking like this [PR
+that adds Playwright.](https://github.com/bokeh/bokeh/pull/14032)
+
+## Progressive work going forward
+
+Going from 0 to 1 was a good start, but to go from 1 to n, there are more things
+that can be done in Bokeh. In the future, we could:
+
+- Add Integration snapshot tests that will be marked as known failures.
+- Add CI checks for accessibility tests. This would prevent commits from being
+  merged if they will change Bokeh data visualizations in a way that triggers
+  accessibility issues.
+- Couple things a little tighter with the existing convention. In Bokeh,
+  Javascript tests are run with a specific type of command: `node make test:suite_name`. There are certain commands for pairing types of tests
+  together, and these commands could be worked upon or modified as needed.
+- Have a labeling system to be able to run only "accessibility-focused" tests at
+  will. In Bokeh, you can run both individual and group tests using a search
+  string command. The syntax is: `node make test:suite_name -k "search string"`.
+  For labeling, you could prepend (or append) a string like
+  "accessibility-focused" to the code for the tests and run the matching search
+  string in the syntax to run all matching tests. This feature becomes even more
+  useful as `unit` and `integration` directories start to contain more
+  accessibility tests.
```

Comment:
It might make more sense to write that right now, in order to run both:

1. The existing tests, and
2. the tests that you added

You have to run two separate commands. So future work could involve making it so that all tests run from a single command. And that one requirement of this single command would be a mechanism to run only the accessibility-related tests.

Reply:
Nice!!